### PR TITLE
add video-text CLIP model builder

### PR DESCRIPTION
**Pull request by sophiazhi**
Summary:
Add video-text CLIP model builder using MUGEN's architecture.

Test plan:
`python -m pytest examples/mugen/test/retrieval/test_video_clip.py::TestVideoCLIPModel -rP`

### CODE REVIEW

File: examples/mugen/retrieval/video_clip.py [hunk 1]
```diff
@@ -72,14 +81,115 @@ class VideoEncoder(nn.Module):
     """Encode videos to a fixed size vector. Adapted from VideoCLIP
```
Comment (langong347) [inline examples/mugen/retrieval/video_clip.py:81]: "Fixed size vector" can be ambiguous. Can we elaborate on what `VideoEncoder` does here?

File: examples/mugen/retrieval/video_clip.py [hunk 2]
```diff
@@ -23,6 +24,8 @@ class TextEncoder(nn.Module):
     Args:
         pretrained (bool): whether to use a pretrained model or not.
             Defaults to True.
+        trainable (bool): whether the encoder's weights should be trainable (not frozen).
+            Defaults to True. Ignored if ``pretrained`` is False.
```
Comment (langong347) [inline examples/mugen/retrieval/video_clip.py:28]: Would add a warning statement for "no-surprise" reason, e.g.,  when a user says `pretrained = False` and `trainable = False` but instead the behavior is `trainable = True`.

File: examples/mugen/retrieval/video_clip.py [hunk 3]
```diff
@@ -72,14 +81,115 @@ class VideoEncoder(nn.Module):
...
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        embed1 = self.linear1(x)
+        embed2 = self.gelu(embed1)
+        embed2 = self.linear2(embed2)
+        embed2 = self.drop(embed2)
+        embeds = self.layer_norm(embed1 + embed2)
+        return embeds
+
+
+def videoclip(
```
Comment (langong347) [inline examples/mugen/retrieval/video_clip.py:141]: `build_videoclip`

File: examples/mugen/retrieval/video_clip.py [hunk 4]
```diff
@@ -72,14 +81,115 @@ class VideoEncoder(nn.Module):
...
+        text_model,
+        Projection(
+            text_model.model.config.dim, out_dim=proj_out_dim, dropout_prob=proj_dropout
+        ),
+    )
+    video_encoder = nn.Sequential(
+        VideoEncoder(
+            trainable=video_trainable,
+        ),
+        Projection(1024, out_dim=proj_out_dim, dropout_prob=proj_dropout),
```
Comment (langong347) [inline examples/mugen/retrieval/video_clip.py:193]: Hard coding intermediate dimensions can be error prone as it assumes what's returned by the previous layer. 
Is `1024` the last dim of `VideoEncoder` (the concat dim of [`Mixed5c`](https://github.com/pytorch/vision/blob/6512146e447b69cc9fb379eb05e447a17d7f6d1c/test/test_datasets_utils.py#L20-L29))? 
What if we create an attribute `VideoEncoder.out_dim` or `S3D.out_dim` so if there's any inconsistency the error will be self-contained within the module that causes the error.

File: examples/mugen/test/retrieval/test_video_clip.py [hunk 5]
```diff
@@ -82,3 +100,62 @@ def test_forward(self, start):
...
+        return proj, input
+
+    def test_forward(self, utils):
+        proj, input = utils
+        out = proj(input)
+        expected = torch.Tensor([[-1.2214, -0.0066, 1.2280], [-1.3886, 0.4626, 0.9260]])
+        assert_expected(actual=out, expected=expected, rtol=0, atol=1e-4)
+
+
+class TestVideoCLIPModel:
```
Comment (langong347) [inline examples/mugen/test/retrieval/test_video_clip.py:128]: nit: `TestVideoCLIPBuilder` ?

File: examples/mugen/test/retrieval/test_video_clip.py [hunk 6]
```diff
@@ -82,3 +100,62 @@ def test_forward(self, start):
...
+                [101, 6315, 3793, 7099, 2005, 5604, 19204, 17629, 102],
+                [101, 2117, 7820, 3793, 102, 0, 0, 0, 0],
+            ]
+        ).to(dtype=int)
+        input_video = torch.randint(10, [2, 3, 32, 32, 32]).float()
+        return input_text, input_video
+
+    def test_forward(self, utils):
+        input_text, input_video = utils
+        clip = videoclip()
```
Comment (langong347) [inline examples/mugen/test/retrieval/test_video_clip.py:146]: nit: `model` instead of `clip`?

File: examples/mugen/test/retrieval/test_video_clip.py [hunk 7]
```diff
@@ -82,3 +100,62 @@ def test_forward(self, start):
...
+        clip_output = clip(features_a=input_text, features_b=input_video)
+        assert_expected(
+            actual=clip_output.embeddings_a.shape, expected=torch.Size([2, 256])
+        )
+        assert_expected(
+            actual=clip_output.embeddings_b.shape, expected=torch.Size([2, 256])
+        )
+
+    def test_text_dim(self, utils):
```
Comment (langong347) [inline examples/mugen/test/retrieval/test_video_clip.py:155]: why not adding `test_video_dim`?

**Review by langong347 — COMMENTED** [review]
I assume this PR will change after #125 . If that's the case, use `ghstack` or make #125 your base commit will include the unmerged changes.

File: examples/mugen/retrieval/video_clip.py [hunk 8]
```diff
@@ -72,14 +81,115 @@ class VideoEncoder(nn.Module):
...
+
     """
 
-    def __init__(self):
+    def __init__(self, trainable: bool = True):
         super().__init__()
         self.model = S3D(400)
         self.model.fc = nn.Identity()
 
+        if not trainable:
```
[the remark this answers is not in the document]
Comment (sophiazhi) [reply examples/mugen/retrieval/video_clip.py:98]: Right, whichever of #124 and #125 lands first, I'll edit the other one so VideoEncoder has trainable and pretrained args, and line 98 will be ``if pretrained and not trainable``

File: examples/mugen/retrieval/video_clip.py (hunk 4 above)
Comment (sophiazhi) [reply examples/mugen/retrieval/video_clip.py:193]: yup, 1024 is the size of the output of Mixed5c. I'll add the out_dim attributes to both encoders

File: examples/mugen/retrieval/video_clip.py (hunk 1 above)
Comment (sophiazhi) [reply examples/mugen/retrieval/video_clip.py:81]: "Encode videos to the last layer before the fully-connected layer of S3D"?

File: examples/mugen/test/retrieval/test_video_clip.py (hunk 7 above)
Comment (sophiazhi) [reply examples/mugen/test/retrieval/test_video_clip.py:155]: I originally added `test_text_dim` because the text encoder output dim is configurable and `build_videoclip` relied on that logic. Now that the output dim logic is contained within the text encoder, I don't think this test case is needed anymore (dimensions are checked in `test_forward` already)

Comment (sophiazhi) [reply examples/mugen/test/retrieval/test_video_clip.py:155]: * to clarify I don't think `test_text_dim` is needed anymore and similarly no `test_video_dim`

**Comment by Justinyu1618** [comment]
lgtm

**Review by ebsmothers — CHANGES_REQUESTED** [review]
How come changes from other PRs are getting picked up here? Maybe you need to rebase your changes on latest main branch

File: examples/mugen/retrieval/video_clip.py [hunk 9]
```diff
@@ -155,13 +169,15 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
         return embeds
 
 
-def videoclip(
+def build_videoclip(
```
Comment (langong347) [inline examples/mugen/retrieval/video_clip.py:172]: Sorry my bad. I checked again that we didn't use `build` prefix in other builders. 

https://github.com/facebookresearch/multimodal/blob/main/torchmultimodal/models/clip.py

File: examples/mugen/test/retrieval/test_video_clip.py [hunk 10]
```diff
@@ -128,7 +129,31 @@ def make_input_video(c_dim=1):
...
+        out = encoder(input_video)
+        expected_sum = 1829.8858
+        assert_expected(
+            actual=out.shape, expected=torch.Size([2, 1024])
+        )  # batch x embedding
+        assert_expected(
+            actual=out.sum(), expected=torch.as_tensor(expected_sum), rtol=0, atol=1e-3
+        )
+        assert next(encoder.parameters()).requires_grad
+        assert encoder.out_dim == 1024
```
Comment (langong347) [inline examples/mugen/test/retrieval/test_video_clip.py:147]: why need to enforce this value here? are there any subsequent routines that depend on out_dim=1024?

File: examples/mugen/retrieval/video_clip.py [hunk 11]
```diff
@@ -76,36 +87,153 @@ def forward(self, input_ids: torch.Tensor) -> torch.Tensor:
...
+        trainable: bool = True,
         pretrain_path: str = PRETRAINED_S3D_KINETICS400_URL,
     ):
         super().__init__()
         self.model = S3D(400)
+        self.out_dim = self.model.fc.in_channels
         self.model.fc = nn.Identity()
 
         if pretrained:
             self.load_model(pretrain_path)
```
Comment (langong347) [inline examples/mugen/retrieval/video_clip.py:123]: nit: Add a prompt here for the user (the same for TextEncoder as for distilledBERT)

https://github.com/mugen-org/MUGEN_baseline/blob/main/lib/models/videoclip/modules.py#L80

File: examples/mugen/retrieval/video_clip.py [hunk 12]
```diff
@@ -76,36 +87,153 @@ def forward(self, input_ids: torch.Tensor) -> torch.Tensor:
...
 
     Inputs:
         x (Tensor): batch of videos with dimensions (batch, channel, time, height, width)
+            Size of ``channel`` dimension must be 3.
+
     """
 
     def __init__(
         self,
         pretrained: bool = True,
```
Comment (langong347) [inline examples/mugen/retrieval/video_clip.py:113]: if pretrained is True, the user likely don't want to train the model. It looks over complicated here. Apart from MUGEN, have you seen elsewhere that requires both flags?

File: examples/mugen/test/retrieval/test_video_clip.py [hunk 13]
```diff
@@ -72,8 +77,14 @@ def test_forward_pretrained(self, utils):
         assert_expected(
             actual=out.sum(), expected=torch.as_tensor(expected_sum), rtol=0, atol=1e-4
         )
+        assert next(encoder.parameters()).requires_grad
```
Comment (langong347) [inline examples/mugen/test/retrieval/test_video_clip.py:80]: Think about whether we indeed need two knobs (pretrained, trainable). Do they really serve the user? If not, we could probably simplify our API as well as reduce complexity in unit tests.

File: examples/mugen/retrieval/video_clip.py [hunk 14]
```diff
@@ -76,36 +87,153 @@ def forward(self, input_ids: torch.Tensor) -> torch.Tensor:
...
+        trainable: bool = True,
         pretrain_path: str = PRETRAINED_S3D_KINETICS400_URL,
     ):
         super().__init__()
         self.model = S3D(400)
+        self.out_dim = self.model.fc.in_channels
         self.model.fc = nn.Identity()
 
         if pretrained:
```
Comment (langong347) [inline examples/mugen/retrieval/video_clip.py:122]: Comment about elevating pretrained related logic to the builder level. See #136

File: examples/mugen/test/retrieval/test_video_clip.py (hunk 10 above)
Comment (sophiazhi) [reply examples/mugen/test/retrieval/test_video_clip.py:147]: wanted to check that the out_dim logic was correct

File: examples/mugen/retrieval/video_clip.py (hunk 12 above)
Comment (sophiazhi) [reply examples/mugen/retrieval/video_clip.py:113]: as far as i know it's common to use a pretrained model and then continue training it, but I'm open to other suggestions for simplifying the API if possible

File: examples/mugen/test/retrieval/test_video_clip.py (hunk 13 above)
Comment (sophiazhi) [reply examples/mugen/test/retrieval/test_video_clip.py:80]: resolving since we can continue discussion in the comment above

**Review by langong347 — APPROVED** [review]
